src/main/scala/com/atomist/rug/runtime/js/JavaScriptInvokingProjectOperation.scala

```diff
@@ -126,17 +128,36 @@ abstract class JavaScriptInvokingProjectOperation(
 
         p.setValidInputDescription(details.get("validInput").asInstanceOf[String])
         p.describedAs(details.get("description").asInstanceOf[String])
-        details.get("pattern").asInstanceOf[String] match {
+
+        pPattern match {
           case s: String if s.startsWith("@") => DefaultIdentifierResolver.resolve(s.substring(1)) match {
             case Left(_) =>
-              throw new InvalidRugParameterPatternException(s"Unable to recognized predefined validation pattern: $s")
+              throw new InvalidRugParameterPatternException(s"Unable to recognize predefined validation pattern for parameter $pName: $s")
             case Right(pat) => p.setPattern(pat)
           }
-          case s: String if !s.startsWith("^") || !s.endsWith("$") => throw new InvalidRugParameterPatternException(s"Parameter $name does not contain anchors: $s")
+          case s: String if !s.startsWith("^") || !s.endsWith("$") =>
+            throw new InvalidRugParameterPatternException(s"Parameter $pName validation pattern must contain anchors: $s")
           case s: String => p.setPattern(s)
-          case _ =>
+          case _ => throw new InvalidRugParameterPatternException(s"Parameter $pName has no valid validation pattern")
+        }
+
+        val avs = details.get("allowed_values").asInstanceOf[ScriptObjectMirror]
+        val allowedValues: Seq[String] =
+          if (avs == null || avs.asScala.isEmpty) Seq.empty
+          else if (avs.isArray) avs.values.toArray.map(av => av.toString).toSeq
+          else Seq.empty
+        p.setAllowedValues(allowedValues.map(av => AllowedValue(av, av)))
```

Comment:
So I left allowed_values out deliberately because I wasn't too sure what it was for, especially given the field names of AllowedValues. Additionally, regular expressions can normally cover most cases adequately....

If name/displayName are meaningful in someway to someone, should we expose them in TS directly? Otherwise, let's just make it a Set of string values (which makes sense), but make it mutually exclusive with pattern.